src/devices.jl

```diff
@@ -0,0 +1,29 @@
+import ClimaComms
+
+# Sometimes, it is convenient to know whether something is running on a device or on its
+# host. We define new AbstractDevices and AbstractCommsContext to identify the case of a
+# function running on the device with data on the device.
+
+"""
+    DeviceSideDevice()
+
+This device represents data defined on the device side of an accelerator.
+
+The most common example is data defined on a GPU. DeviceSideDevice() is used for
+operations within the accelerator.
+"""
+struct DeviceSideDevice <: ClimaComms.AbstractDevice end
```

Comment:
I've thought a bit more about this: why don't we put this into ClimaComms?

Reply:
This would be a device that behaves differently compared to the other ones in ClimaComms. It doesn't have any associated features, and it cannot be "selected". It is just a tag at the end of the day. 

But I don't have a very strong opinion about it.